subsys/bluetooth/mesh/provisioner.c

```diff
@@ -751,17 +751,16 @@ int bt_mesh_auth_method_set_output(bt_mesh_output_action_t action, uint8_t size)
 
 int bt_mesh_auth_method_set_static(const uint8_t *static_val, uint8_t size)
 {
-	uint8_t tail_size = size < PROV_AUTH_MAX_LEN ? PROV_AUTH_MAX_LEN - size : 0;
-
 	if (!size || !static_val) {
 		return -EINVAL;
 	}
 
 	prov_set_method(AUTH_METHOD_STATIC, 0, 0);
 
-	memcpy(bt_mesh_prov_link.auth + tail_size, static_val,
-	       tail_size ? size : PROV_AUTH_MAX_LEN);
-	memset(bt_mesh_prov_link.auth, 0, tail_size);
+	memset(bt_mesh_prov_link.auth, 0, PROV_AUTH_MAX_LEN);
+	/* Trim the Auth if it is longer than required length */
+	memcpy(bt_mesh_prov_link.auth, static_val,
```

Comment:
Wrong zero padding.

Reply:
if we take the example in document:
--------------------------
lets:
PROV_AUTH_MAX_LEN=16
static_auth = [0x12, 0x34, 0x56]
size = 3

what used to happen
------------------------
with `tail_size = 13 // 16 - 3`
after
`memcpy(bt_mesh_prov_link.auth + tail_size, static_val, tail_size ? size : PROV_AUTH_MAX_LEN);`
bt_mesh_prov_link.auth = [X, X, X, X, X, X, X, X, X, X, X, X, X, X, 0x12, 0x34, 0x56]

and then
`memset(bt_mesh_prov_link.auth, 0, tail_size);`
bt_mesh_prov_link.auth = [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x12, 0x34, 0x56]

So it works only if the size = PROV_AUTH_MAX_LEN since there is no padding

with the change:
----------------------
`memset(bt_mesh_prov_link.auth, 0, PROV_AUTH_MAX_LEN);` zero all values since we end up padding it with zeros anyway.
`memcpy(bt_mesh_prov_link.auth, static_val, size > PROV_AUTH_MAX_LEN ? PROV_AUTH_MAX_LEN : size);`
bt_mesh_prov_link.auth = [0x12, 0x34, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]

another case:
static_auth = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19]
size = 20
bt_mesh_prov_link.auth = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]

Reply:
> ## if we take the example in document:
> lets: PROV_AUTH_MAX_LEN=16 static_auth = [0x12, 0x34, 0x56] size = 3
> 
> ## what used to happen
> with `tail_size = 13 // 16 - 3` after `memcpy(bt_mesh_prov_link.auth + tail_size, static_val, tail_size ? size : PROV_AUTH_MAX_LEN);` bt_mesh_prov_link.auth = [X, X, X, X, X, X, X, X, X, X, X, X, X, X, 0x12, 0x34, 0x56]
> 
> and then `memset(bt_mesh_prov_link.auth, 0, tail_size);` bt_mesh_prov_link.auth = [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x12, 0x34, 0x56]
> 

Isn't this how specification mandates?

> So it works only if the size = PROV_AUTH_MAX_LEN since there is no padding
> 

Didn't get this. What is changed if size 15, 14 etc to 1? 

> ## with the change:
> `memset(bt_mesh_prov_link.auth, 0, PROV_AUTH_MAX_LEN);` zero all values since we end up padding it with zeros anyway. `memcpy(bt_mesh_prov_link.auth, static_val, size > PROV_AUTH_MAX_LEN ? PROV_AUTH_MAX_LEN : size);` bt_mesh_prov_link.auth = [0x12, 0x34, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
> 

Please take a look at specification 5.4.2.4 Authentication(especially samples). You did it in the opposite manner to what the specification says.

> another case: static_auth = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19] size = 20 bt_mesh_prov_link.auth = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]

I'm pretty sure the current implementation does everything correctly. Oppositely, this PR breaks OOB static authentication completely.

Reply:
Seems, @m-alperen-sener, you are right. It was already reported here once: https://github.com/zephyrproject-rtos/zephyr/issues/52169 and fixed here: https://github.com/zephyrproject-rtos/zephyr/pull/52255
but in some moment all these changes have been lost. Probably during Mesh-1.1 merge.

Reply:
I guess, it is better to apply my previous changes first and then add your modifications. The current changes in the worst case make a copy up to 64 bytes (considering initial zeroing). 
In my previous PR, I refused it.